### PR TITLE
ddl: fix a bug that MDL may progress unexpectedly or block forever

### LINKED ISSUE #46920: DDL maybe block forever is owner is restart or progress the schema change unexpectedly
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Given two tables, namely t1, t2.
And two TiDB servers, tidb1(ddl owner), tidb2.
Note: It may not reproduce for 2 tables/DDLs, you can extend to 3 tables/DDLs.
1. connect to tidb 2. Do:
``` sql
begin;
select * from t1, t2;
```
2. connect to tidb 2. Do:
``` sql
alter table t1 add column b int;
```
3. connect to tidb 2. Do:
``` sql
alter table t2 add column b int;
```
4. restart/close/kill TiDB1(owner)
5. commit the transaction in step1 session.

### 2. What did you expect to see? (Required)
Two DDL success finally.

### 3. What did you see instead (Required)
The first DDL success finally, but the second DDL block forever(even with restarting the whole cluster),

### 4. What is your TiDB version? (Required)
6.5, (7.1, master) No blocking, but progress the schema change unexpectedly .

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #46920

Problem Summary:

### What is changed and how it works?
1. Check once for every DDL job.
2. Check if waitSchemaChange is returned correctly.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
Pass the test described in the issue
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix a bug that MDL may progress unexpectedly or block forever
```

### CODE REVIEW

File: ddl/job_table.go [hunk 1]
```diff
@@ -378,7 +378,7 @@ func (d *ddl) delivery2worker(wk *worker, pool *workerPool, job *model.Job) {
 			metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Dec()
 		}()
 		// check if this ddl job is synced to all servers.
-		if !d.isSynced(job) || d.once.Load() {
+		if !d.isSynced(job) || !d.maybeAlreadyRunOnce(job.ID) {
```
[the remark this answers is not in the document]
Comment (wjhuang2016) [reply ddl/job_table.go:381]: After GC the onceMap, the jobs need to do an extra checkMDLInfo. If waitSchemaSynced passed already, checkMDLInfo would return nil for this job.

File: session/bootstraptest/bootstrap_upgrade_test.go [hunk 2]
```diff
@@ -586,13 +586,6 @@ func TestUpgradeVersionForResumeJob(t *testing.T) {
 			jobID = job.ID
 			times = 1
 		}
-		// Make sure we do jobID first, then do jobID+1.
-		if job.ID == jobID && job.SchemaState == model.StateWriteReorganization && job.State == model.JobStateQueueing && times == 1 {
-			times = 2
-		}
-		if job.ID == jobID+1 && job.SchemaState == model.StateNone && job.State == model.JobStateQueueing && times == 2 {
-			times = 3
-		}
```
Comment (tangenta) [inline session/bootstraptest/bootstrap_upgrade_test.go:595]: Why do you remove it?

**Review by tangenta — APPROVED** [review]

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -842,11 +842,11 @@ func (w *worker) HandleDDLJobTable(d *ddlCtx, job *model.Job) (int64, error) {
 			// which may act like a deadlock.
 			w.jobLogger(job).Info("run DDL job failed, sleeps a while then retries it.",
 				zap.Duration("waitTime", GetWaitTimeWhenErrorOccurred()), zap.Error(runJobErr))
-		}
 
-		// In test and job is cancelling we can ignore the sleep
-		if !(intest.InTest && job.IsCancelling()) {
-			time.Sleep(GetWaitTimeWhenErrorOccurred())
+			// In test and job is cancelling we can ignore the sleep
```
Comment (zimulala) [inline ddl/ddl_worker.go:846]: Could we combine conditions line838 and line840?

**Comment by wjhuang2016** [comment]
/test unit-test


<br class="Apple-interchange-newline">

**Pull request merged**